### ISSUE TITLE
parquet_derive does not need to depend on parquet

### ISSUE THREAD

**Issue by mdekstrand** [apache/arrow-rs#1002]
**Describe the bug**

Currently, `parquet_derive` depends on `parquet`.

I do not think this is necessary, and it seems to be causing Cargo to build Arrow twice: once for use in the macro, and again for use in the program itself.

**To Reproduce**

- Depend on `parquet` and `parquet_derive` in a program.
- Run `cargo build`
- See that it lists `parquet` twice in the progress bar, and that two compiled copies are in `target/$PROFILE/deps`

**Expected behavior**

`parquet`, `arrow`, and friends to only be compiled once, and only appear once in `target/$PROFILE/deps`.

**Additional context**

**Comment by ByteBaker** [comment]
@tustvold @alamb is this worth doing something? I have a solution which involves putting local types inside `parquet_derive` rather than those imported from `parquet` but not sure if it's the way to go.

**Comment by alamb** [comment]
This bug report seems strange to me -- I think having 2 versions of parquet doesn't sound right -- maybe the first thing to do is to reproduce the report (it is quite old now)

If it is indeed a problem then we can consider fixing.

**Comment by mdekstrand** [comment]
@alamb The `parquet_derive` Cargo manifest still lists `parquet` as a dependency: https://github.com/apache/arrow-rs/blob/52920f946cf8aa038d04fe432e77889f8253b7ce/parquet_derive/Cargo.toml#L34-L38

I also re-built my project with the current `parquet_derive` as well as upgraded Rust (in case there was a relevant Rust / Cargo change), and I am no longer seeing `parquet` show up twice in the compiling list, but still seeing two compiled versions of Parquet under `target/debug/deps`. I'm not sure the extent to which this is a feature-selection issue: the `parquet_derive` crate depends on the `parquet` crate with default features, and I might not have been using default features for my own `parquet` dependency when I originally filed the bug. The default-feature dependency for `parquet_derive` also means that the macro package also pulls in Arrow as a dependency.

In any case, I think it would be helpful to review whether `parquet_derive` actually needs to depend on `parquet`. Since macros operate at a purely syntactic level, it is very possible that the derive macros can do everything they need without an actual `parquet` dependency. If it still does need the `parquet` dependency, does it need the full default feature set including Arrow?

The fundamental reason that this can wind up with double-compilation is that macros and the underlying code are compiled for two different targets / environments: the macro is compiled to be loaded into the Rust interpreter (and therefore in a cross-compilation scenario, which I am not in, would be compiled for the build host rather than the target system), and used to generate code that will be compiled for the target system.

**Comment by alamb** [comment]
> In any case, I think it would be helpful to review whether parquet_derive actually needs to depend on parquet. 

Doesn't parquet_derive macros add calls to things in the parquet crate? I don't understand how it would work to not depend on parquet (though maybe a PR would make it clearer)

>  but still seeing two compiled versions of Parquet under target/debug/deps. I'm not sure the extent to which this is a feature-selection issue: 

I suspect you are right it is a feature selection issue (a feature of rust, rather than anything specific with this crate)

**Comment by mdekstrand** [comment]
> Doesn't parquet_derive macros add calls to things in the parquet crate? I don't understand how it would work to not depend on parquet (though maybe a PR would make it clearer)

Yes, but the macro doesn't need to depend on the crate in order to do that. Macros just emit Rust syntax tokens, which the compiler then interprets in the contexts of the crates being linked into the program that uses the macro. Any program *using* `parquet_derive` needs to depend on `parquet` (and it already needs to, because transitive dependencies aren't available to use directly in a crate), but macros don't need to depend on things they generate code for.

**Comment by alamb** [comment]
Is there precident for macros in crates that generate calls to crates they
don't also depend on ? How would we ensure that parquet_derive is calling
into supported APIs for the parquet crate?

**Comment by mdekstrand** [comment]
Serde's derive crate does not depend on the main crate: https://docs.rs/serde_derive/latest/serde_derive/

In fact, the main crate depends on the derive crate (when the `"derive"` feature is activated), so that client code can simply depend on `serde` with the `"derive"` feature to get the derivation macros without needing to depend on another crate.

As I understand the Rust macro system, depending on the crate doesn't necessarily ensure you're calling supported APIs anyway - since macros just emit tokens, AFAIK there isn't anything that checks the validity until the compiler is actually compiling the program that *uses* the macro.

**Comment by alamb** [comment]
Sounds like a good prior art -- maybe time to make a PR
